src/app/backend/userlinks/userlinks.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2017 The Kubernetes Dashboard Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package userlinks
+
+import (
+	"encoding/json"
+	"log"
+	"strconv"
+
+	"strings"
+
+	"github.com/kubernetes/dashboard/src/app/backend/api"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sClient "k8s.io/client-go/kubernetes"
+)
+
+const (
+	annotationObj     = "alpha.dashboard.kubernetes.io/links"
+	apiserverProxyURL = "http://{{apiserver-proxy-url}}"
+)
+
+// UserLink is an optional annotation attached to the pod or service resource objects.
+type UserLink struct {
+	// Description is the key specified in the annotationObj set
+	Description string `json:"description"`
+	// Link is the value specified in the annotationObj set
+	Link string `json:"link"`
+	// Link status
+	Valid bool `json:"valid"`
+}
+
+// GetUserLinks delagates getting of user links based on passed in resource kind (ResourceKindService, ResourceKindPod)
+func GetUserLinks(client k8sClient.Interface, namespace, name, resource, host string) (userLinks []UserLink, err error) {
+	log.Printf("Getting %s resource in %s namespace", name, namespace)
+
+	switch {
+	case resource == api.ResourceKindService:
+		return getServiceLinks(client, namespace, name, host)
+	case resource == api.ResourceKindPod:
+		return getPodLinks(client, namespace, name, host)
+	default:
+		log.Printf("Unknown resource types %T!\n", resource)
+	}
+	return
+}
+
+// getServiceLinks get userlinks for services
+func getServiceLinks(client k8sClient.Interface, namespace, name, host string) ([]UserLink, error) {
+	userLinks := []UserLink{}
+	service, err := client.CoreV1().Services(namespace).Get(name, metaV1.GetOptions{})
+	if err != nil {
+		return userLinks, err
+	}
+	if len(service.Annotations[annotationObj]) == 0 {
+		return userLinks, err
+	}
```

Comment:
Merge both ifs to err != nil || len(pod.Annotations[annotationObj]) == 0.

Reply:
`getServiceLinks` and `getPodLinks` have a lot in common. Perhaps you could avoid such code duplication?

Reply:
I tried but they differ enough so that it makes it not worth refactoring further.